src/main/java/uk/gov/hmcts/reform/managecase/service/noc/PrepareNoCService.java

```diff
@@ -0,0 +1,197 @@
+package uk.gov.hmcts.reform.managecase.service.noc;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.fasterxml.jackson.databind.node.TextNode;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import uk.gov.hmcts.reform.idam.client.models.UserInfo;
+import uk.gov.hmcts.reform.managecase.client.datastore.CaseDetails;
+import uk.gov.hmcts.reform.managecase.client.definitionstore.model.CaseRole;
+import uk.gov.hmcts.reform.managecase.client.prd.FindUsersByOrganisationResponse;
+import uk.gov.hmcts.reform.managecase.domain.DynamicListElement;
+import uk.gov.hmcts.reform.managecase.domain.OrganisationPolicy;
+import uk.gov.hmcts.reform.managecase.repository.DefinitionStoreRepository;
+import uk.gov.hmcts.reform.managecase.repository.PrdRepository;
+import uk.gov.hmcts.reform.managecase.security.SecurityUtils;
+import uk.gov.hmcts.reform.managecase.util.JacksonUtils;
+
+import javax.validation.ValidationException;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static java.lang.String.format;
+import static java.util.stream.Collectors.toList;
+import static org.apache.commons.lang3.StringUtils.isBlank;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+import static uk.gov.hmcts.reform.managecase.api.errorhandling.ValidationError.CASE_TYPE_CANNOT_BE_BLANK;
+import static uk.gov.hmcts.reform.managecase.api.errorhandling.ValidationError.JURISDICTION_CANNOT_BE_BLANK;
+import static uk.gov.hmcts.reform.managecase.api.errorhandling.ValidationError.MISSING_COR_ON_THE_CASE_DATA;
+import static uk.gov.hmcts.reform.managecase.api.errorhandling.ValidationError.NO_ORGANISATION_ID_IN_ANY_ORG_POLICY;
+import static uk.gov.hmcts.reform.managecase.api.errorhandling.ValidationError.NO_ORGANISATION_POLICY_ON_CASE_DATA;
+import static uk.gov.hmcts.reform.managecase.api.errorhandling.ValidationError.NO_SOLICITOR_ORGANISATION_RECORDED_IN_ORG_POLICY;
+import static uk.gov.hmcts.reform.managecase.api.errorhandling.ValidationError.ONGOING_NOC_REQUEST;
+import static uk.gov.hmcts.reform.managecase.api.errorhandling.ValidationError.ORG_POLICY_CASE_ROLE_NOT_IN_CASE_DEFINITION;
+
+@Service
+@SuppressWarnings({"PMD.UseLocaleWithCaseConversions"})
+public class PrepareNoCService {
+    protected static final String COR_CASE_ROLE_ID = "CaseRoleId";
+    protected static final String COR_REQUEST_TIMESTAMP = "RequestTimestamp";
```

Comment:
If these field IDs are being referenced in the client data models (see _src/main/java/uk/gov/hmcts/reform/managecase/client/datastore/CaseDetails.java_) can they not be imported instead of redefined?